orchagent/copporch.cpp

```diff
@@ -0,0 +1,479 @@
+                            return task_process_status::task_failed;
+                        }                    
+                    }
+                }
+            }
+            SWSS_LOG_DEBUG("Applying trap group attributes");
+            for(sai_uint32_t ind = 0; ind < trap_gr_attribs.size(); ind++)
+            {
+                auto trap_gr_attr = trap_gr_attribs[ind];
+                SWSS_LOG_DEBUG("Applying trap group attribute:%d", trap_gr_attr.id);
+                if (SAI_STATUS_SUCCESS != (sai_status = sai_hostif_api->set_trap_group_attribute(m_trap_map[key], &trap_gr_attr)))
+                {
+                    SWSS_LOG_ERROR("Failed to apply attribute:%d to trap group:%llx, name:%s, error:%d\n", trap_gr_attr.id, m_trap_map[key], key.c_str(), sai_status);
+                    return task_process_status::task_failed;
+                }
+            }                
+        }
+        else
+        {
+            SWSS_LOG_DEBUG("Creating new trap_group object:%s", key.c_str());
+            sai_object_id_t new_trap;
+            if (SAI_STATUS_SUCCESS != (sai_status = sai_hostif_api->create_hostif_trap_group(&new_trap, trap_gr_attribs.size(), trap_gr_attribs.data())))
+            {
+                SWSS_LOG_ERROR("Failed to create new trap_group with name:%s", key.c_str());
+                return task_process_status::task_failed;
+            }
+            SWSS_LOG_DEBUG("Created new trap_group:%llx with name:%s", new_trap, key.c_str());
+            m_trap_map[key] = new_trap;
+
+            if(!policer_attribs.empty())
+            {
+                if (SAI_STATUS_SUCCESS != (sai_status = sai_policer_api->create_policer(&policer_id, policer_attribs.size(), policer_attribs.data())))
+                {
+                    SWSS_LOG_ERROR("Failed to create policer for new trap_group:%llx, name:%s, error:%d", new_trap, key.c_str(), sai_status);
+                    return task_process_status::task_failed;
+                }
+                SWSS_LOG_DEBUG("Created new policer:%llx", policer_id);
+                attr.id = SAI_HOSTIF_TRAP_GROUP_ATTR_POLICER;
+                attr.value.oid = policer_id;
+                if (SAI_STATUS_SUCCESS != (sai_status = sai_hostif_api->set_trap_group_attribute(new_trap, &attr))) {
+                    SWSS_LOG_ERROR("Failed to bind policer:%llx to trap gorup:%llx, name:%s, error:%d\n", policer_id, new_trap, key.c_str(), sai_status);
+                    return task_process_status::task_failed;
+                }
+                SWSS_LOG_DEBUG("Bound policer to the trap group");
+            }
+        }
+
+        //
+        // The channel is always the NETDEV device.
+        //
+        attr.id = SAI_HOSTIF_TRAP_ATTR_TRAP_CHANNEL;
+        attr.value.s32 = SAI_HOSTIF_TRAP_CHANNEL_NETDEV;
+        trap_id_attribs.push_back(attr);
```

Comment:
>trap_id_attribs [](start = 8, length = 15)

you are building trap_id_attribs in multiple places. 

Can you consolidate the attributes building in one place. It's easier for readability. 

here is my suggestion:

1. get key values from db, fill your own trap object.

e.g, map<trap_id, tuple<trap_group_name, trap_action>>

2. In one function, translate the trap object into sai trap object and call corresponding sai function.

set trap_id 
set trap_channel
set trap_group
call set_trap_attribute()